File: packages/skill-registry/typescript/src/manifest.ts
/**
 * SkillManifest — the canonical description of a skill in the axis ecosystem.
 *
 * It is the EXISTING `SKILL.md` frontmatter (name, description, tools_exposed,
 * platforms, user-invocable) EXTENDED with the fields needed to route a skill to
 * the right transport: `type`, `version`, `io_schema`, `transport`, `dependencies`,
 * `tags`, `tenants`. All new fields are optional so the 19 axis-runtime SKILL.md
 * files keep parsing unchanged; `type` is inferred when absent.
 *
 * Decided in Phase 5 (05-SKILL-CONTRACT-SPEC): we extend SKILL.md rather than invent
 * a separate `skill.yaml` — one format, already adopted.
 */

/** Current manifest format version. Loaders reject unknown versions. */
export const MANIFEST_VERSION = 1 as const;

/**
 * The three kinds of skill (see the SPEC). Routing depends on this:
 *  - tool        → a callable capability, usually served via MCP (shared, cross-agent)
 *  - inprocess   → hot-path / stateful logic that stays a local function in the agent
 *  - declarative → procedural know-how injected into the LLM system prompt (Agent Skills)
 */
export type SkillType = "tool" | "inprocess" | "declarative";

export type SkillTransport =
  | { kind: "mcp"; server: string; tool: string }
  | { kind: "inproc"; module: string; export?: string }
  | { kind: "registry"; tool: string };

export interface SkillIoSchema {
  /** JSON-Schema-ish description of the input args (loose by design at v1). */
  input?: Record<string, unknown>;
  /** JSON-Schema-ish description of the output. */
  output?: Record<string, unknown>;
}

export interface SkillManifest {
  // ─── existing SKILL.md frontmatter (unchanged) ──────────────────────────────
  /** kebab-case id, unique within scope. */
  name: string;
  /** When to invoke + what it does (drives LLM tool selection). */
  description: string;
  /** Tool names this skill exposes in the host REGISTRY (axis-runtime convention). */
  tools_exposed?: string[];
  /** OS platforms the skill supports. */
  platforms?: string[];
  /** Whether a user can invoke it directly (Claude Code skills). */
  user_invocable?: boolean;

  // ─── new in the extended manifest (all optional, back-compatible) ───────────
  /** Semver of THIS skill. Defaults to "0.1.0" when absent. */
  version?: string;
  /** Skill kind. Inferred via `inferType()` when absent. */
  type?: SkillType;
  io_schema?: SkillIoSchema;
  /** How the skill resolves at runtime. Required-ish for `type: tool`. */
  transport?: SkillTransport;
  /** Names of other skills this one needs (the AGENT orchestrates; no nested MCP). */
  dependencies?: string[];
  tags?: string[];
  /** Tenants allowed to use this skill (enforced at the MCP server via token). */
  tenants?: string[];
}

/**
 * Infer the skill type when the manifest doesn't state it, so legacy SKILL.md files
 * classify correctly without edits:
 *   - has `transport.kind: mcp`           → tool
 *   - has `tools_exposed`                 → tool
 *   - otherwise                           → declarative
 * (`inprocess` must be declared explicitly — it can't be inferred from frontmatter.)
 */
export function inferType(m: SkillManifest): SkillType {
  if (m.type) return m.type;
  if (m.transport?.kind === "mcp") return "tool";
  if (m.tools_exposed && m.tools_exposed.length > 0) return "tool";
  return "declarative";
}

/** The version of a manifest, defaulting to "0.1.0" when unset. */
export function manifestVersion(m: SkillManifest): string {
  return m.version ?? "0.1.0";
}

File: packages/skill-registry/python/axis_skill_registry/manifest.py
"""Pydantic v2 mirror of the canonical SkillManifest contract.

This module is a faithful, 1:1 mirror of the TypeScript `SkillManifest`
interface (`typescript/src/manifest.ts`) and the JSON Schema
(`skill-manifest.schema.json`). Every field, its optionality, and every
enum/literal is mapped exactly.

`additionalProperties: true` in the schema is mirrored with
`model_config = ConfigDict(extra="allow")` on the root model so that legacy
SKILL.md frontmatter keys (e.g. `allowed-tools`) are tolerated without
validation errors. Optional fields (the TS `?` modifier) are typed
`Optional[...] = None`.

Discriminated unions (`SkillTransport`) are modelled with a pydantic v2
tagged union keyed on the literal `kind` field.
"""
from __future__ import annotations

from typing import Annotated, Any, Dict, List, Literal, Optional, Union

from pydantic import BaseModel, ConfigDict, Field

# ─── transport (discriminated union on `kind`) ────────────────────────────────


class McpTransport(BaseModel):
    """MCP server transport."""

    model_config = ConfigDict(extra="forbid")

    kind: Literal["mcp"]
    server: str
    tool: str


class InprocTransport(BaseModel):
    """In-process (hot-path) transport."""

    model_config = ConfigDict(extra="forbid")

    kind: Literal["inproc"]
    module: str
    export: Optional[str] = None


class RegistryTransport(BaseModel):
    """Registry-routed tool transport."""

    model_config = ConfigDict(extra="forbid")

    kind: Literal["registry"]
    tool: str


SkillTransport = Annotated[
    Union[McpTransport, InprocTransport, RegistryTransport],
    Field(discriminator="kind"),
]

# ─── io schema ────────────────────────────────────────────────────────────────


class SkillIoSchema(BaseModel):
    """JSON-Schema-ish description of skill input/output (loose by design at v1)."""

    model_config = ConfigDict(extra="forbid")

    input: Optional[Dict[str, Any]] = None
    output: Optional[Dict[str, Any]] = None


# ─── root manifest ────────────────────────────────────────────────────────────

SkillType = Literal["tool", "inprocess", "declarative"]

# Current manifest format version. Loaders reject unknown versions.
MANIFEST_VERSION = 1


class SkillManifest(BaseModel):
    """Extended SKILL.md frontmatter.

    Mirror of @axis/skill-registry SkillManifest (v1).
    extra="allow" is intentional — tolerates legacy frontmatter keys such as
    `allowed-tools` without validation errors (additionalProperties:true in
    the JSON Schema).
    """

    model_config = ConfigDict(extra="allow")

    # ─── existing SKILL.md frontmatter (unchanged) ───────────────────────────
    # kebab-case id, unique within scope.
    name: str
    # When to invoke + what it does (drives LLM tool selection).
    description: str
    # Tool names this skill exposes in the host REGISTRY (axis-runtime convention).
    tools_exposed: Optional[List[str]] = None
    # OS platforms the skill supports.
    platforms: Optional[List[str]] = None
    # Whether a user can invoke it directly (Claude Code skills).
    user_invocable: Optional[bool] = None

    # ─── new in the extended manifest (all optional, back-compatible) ────────
    # Semver of THIS skill. Defaults to "0.1.0" when absent.
    version: Optional[str] = None
    # Skill kind. Inferred via `infer_type()` when absent.
    type: Optional[SkillType] = None
    io_schema: Optional[SkillIoSchema] = None
    # How the skill resolves at runtime. Required-ish for `type: tool`.
    transport: Optional[SkillTransport] = None
    # Names of other skills this one needs.
    dependencies: Optional[List[str]] = None
    tags: Optional[List[str]] = None
    # Tenants allowed to use this skill (enforced at the MCP server via token).
    tenants: Optional[List[str]] = None


def infer_type(m: SkillManifest) -> SkillType:
    """Infer the skill type when the manifest doesn't state it.

    Rules (identical to the TypeScript inferType function):
      - has ``transport.kind == "mcp"``       → tool
      - has non-empty ``tools_exposed``        → tool
      - otherwise                              → declarative

    Note: ``inprocess`` must be declared explicitly — it cannot be inferred
    from frontmatter alone.
    """
    if m.type is not None:
        return m.type
    if m.transport is not None and m.transport.kind == "mcp":
        return "tool"
    if m.tools_exposed and len(m.tools_exposed) > 0:
        return "tool"
    return "declarative"


def manifest_version(m: SkillManifest) -> str:
    """Return the version of a manifest, defaulting to '0.1.0' when unset."""
    return m.version if m.version is not None else "0.1.0"
